### PR TITLE
[KSM] Fix `labelsAsTags`

### PR DESCRIPTION
**Pull request by L3n41c**
### What does this PR do?

This is a redesign and rewrite of #13973.

When `labels_as_tags` contained something like:
```yaml
labels_as_tags:
  node:
    node_label: node_tag
  namespace:
    ns_label: ns_tag
```

The node metrics, generated from nodes with a `node_label` label were properly tagged with `node_tag`.
But, pods metrics, generated from pods running inside a namespace with a `ns_label` label were not properly tagged with `ns_tag`.

There was a bug in the tag naming when the monitored resource and the source of the tag were of two different kind.

### Motivation

Makes things work as expected.

### Additional Notes

What the bug shown is that the current approach which consists in splitting the metrics processing in two steps:
* first, getting the prometheus labels to use with the “label joins” algorithm and then
* second, apply a “label mapping” to turn prometheus labels into datadog tags

cannot work when the kind of the resource the label comes from and the kind of the decorated resource are not the same.

So, the configuration of the “label joins” algorithm had to be modified to also embed the remapping part.

Concretely, what this means is that the “label joins” algorithm now need a configuration that looks like:
```go
type joinsConfig struct {
	labelsToMatch []string
	labelsToGet   map[string]string
	getAllLabels  bool
}
```

instead of:
```go
type JoinsConfig struct {
	// LabelsToMatch contains the labels that must
	// match the labels of the targeted metric
	LabelsToMatch []string `yaml:"labels_to_match"`

	// LabelsToGet contains the labels we want to get from the targeted metric
	LabelsToGet []string `yaml:"labels_to_get"`

	// GetAllLabels replaces LabelsToGet if enabled
	GetAllLabels bool `yaml:"get_all_labels"`
}
```

Note that `labelsToGet` became a map (of correspondance between a Prometheus label and a Datadog tag) whereas it used to be a slice (because the mapping was in the `LabelsMapper` map.

However, the old structure (with `LabelsToGet` as a slice) is what is exposed in the user-defined configuration of the check.
So, for backward compatibility reasons, it’s not possible to modify it.
But in order to make things slightly more explicit, the internal GO type has been renamed `JoinsConfigWithoutLabelsMapping`.

So, now, inside the `KSMConfig struct`,
* the only field used by the “label joins” algorithm is the `labelJoins map[string]*joinsConfig` field. But this field is private and cannot be set by the end user. Instead, this private field is set from the 
* `LabelJoins map[string]*JoinsConfigWithoutLabelsMapping 'yaml:"label_joins"'` user visible field thanks to the `(k *KSMCheck) processLabelJoins()` function
* `LabelsAsTags map[string]map[string]string 'yaml:"labels_as_tags"'` user visible field thanks to the `(k *KSMCheck) processLabelsAsTags()` function and
* `AnnotationsAsTags map[string]map[string]string 'yaml:"annotations_as_tags"'` user visible field thanks to the `(k *KSMCheck) processAnnotationsAsTags()` function

### Describe how to test/QA your changes

Create a KSM config with:
```yaml
labels_as_tags:
  node:
    my_label: node_tag
  deployment:
    my_label: deployment_tag
  namespace:
    ns_label: ns_tag
```
Then, label some nodes with `my_label`, some deployments with `my_label` (the same label) and validate that the corresponding metrics have the proper tags (`node_tag` for the former and `deployment_tag` for the latter)

Label some namespaces with `ns_label` and validate that the metrics for the deployment, replicaset and pods running inside this namespace are properly tagged with `ns_tag`.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: releasenotes/notes/ksm_fix_labels_as_tags-2d0c00093f944e9f.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
+  - |
+    Fixes the `labelsAsTags` parameter of the kube-state metrics core check.
+    Tags were not properly formatted when they came from a label on one resource kind (ex. namespace) and turned into a tag on another resource kind (ex. pod).
```
Comment (buraizu) [inline releasenotes/notes/ksm_fix_labels_as_tags-2d0c00093f944e9f.yaml:12]:
```suggestion
    Fixes the `labelsAsTags` parameter of the kube-state metrics core check.
    Tags were not properly formatted when they came from a label on one resource type (for example, namespace) and turned into a tag on another resource type (for example, pod).
```

**Review by buraizu — APPROVED** [review]
This looks good to me and I'm approving, but left some minor suggestions.

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state.go [hunk 2]
```diff
@@ -77,26 +78,26 @@ type KSMConfig struct {
...
 	//       - label_addonmanager_kubernetes_io_mode
-	LabelJoins map[string]*JoinsConfig `yaml:"label_joins"`
+	LabelJoins map[string]*JoinsConfigWithoutLabelsMapping `yaml:"label_joins"`
 
 	// LabelsAsTags
 	// Example:
 	// labels_as_tags:
 	//   pod:
-	//     - "app_*"
+	//     app: pod_app
```
Comment (clamoriniere) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state.go:87]: is it backward compatible?

Comment (clamoriniere) [reply pkg/collector/corechecks/cluster/ksm/kubernetes_state.go:87]: the example was wrong before 👍

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state.go [hunk 3]
```diff
@@ -141,6 +133,9 @@ type KSMConfig struct {
 	// LeaderSkip forces ignoring the leader election when running the check
 	// Can be useful when running the check as cluster check
 	LeaderSkip bool `yaml:"skip_leader_election"`
+
+	// Private field containing the label joins configuration built from `LabelJoins`, `LabelsAsTags` and `AnnotationsAsTags`.
```
Comment (clamoriniere) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state.go:137]: maybe add as comment what is the map key. from what I understand it the the metrics family name.

**Review by clamoriniere — APPROVED** [review]

**Pull request merged**